### ISSUE TITLE
Request: Add "must have side-effects" as an option to "no-sequences"

### ISSUE THREAD

**Issue by MarkKahn** [eslint/eslint#10553]
**What rule do you want to change?**

no-sequences

**Does this change cause the rule to produce more or fewer warnings?**

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

new option

**Please provide some example code that this change will affect:**

```js
(doSomething(), !!test); // valid
(1, 2); // invalid, no side-effects
```

**What does the rule currently do for this code?**

throws regardless of content

**What will the rule do after it's changed?**

Only throw if a sequence is missing side effects, which 100% means it's either useless or likely a typo

**Comment by not-an-aardvark** [comment]
Hi, thanks for the proposal. I'm not sure this is possible to implement because ESLint can't determine whether any given expression has side-effects. There are some expressions that clearly have no side-effects (e.g. number literals), but if the expression involves accessing any properties or calling any functions, then there's no way to tell what code will be invoked and whether it will have any side-effects. For example:

```js
var foo = { bar: 1 };

doSomething(foo);

(foo.bar, !!test); // does this have a side-effect?
```

Normally, the above code would not be expected to have a side-effect, but it will have a side-effect if `doSomething` replaced the `bar` property with a getter that modifies some state.

**Comment by platinumazure** [comment]
Could we use the same heuristic we currently use for `no-unused-expressions` for this enhancement proposal?
